### DISCUSSION TITLE
native snippet nested placeholders

### DISCUSSION THREAD

**Discussion by Old-Farmer** [LazyVim/LazyVim#3322] [Q&A]
Native snippet hasn't supported nested placeholders yet. How to solve this problem? I have really no idea. Back to luasnip?

**Comment by folke** [comment]
They don't support that yet indeed.
There's some code in LazyVim that removes the nesting for those snippets so they would still partially work.
Without that code, `vim.snippet` just errors.

If you need the nesting then luasnip is your only option.

**Reply by Old-Farmer** [reply]
Thanks!

**Reply by kiryph** [reply]
> If you need the nesting then luasnip is your only option.

What about mini.snippets as another alternative? Mini.snippets is in Extras (https://www.lazyvim.org/extras/coding/mini-snippets) and supports nested placeholders:

https://github.com/nvim-mini/mini.nvim/blob/main/readmes/mini-snippets.md#syntax

**Comment by dpetka2001** [comment]
@folke Just so I don't open a new Discussion/Issue, one thing I noticed with native snippets, is that when you accept a suggestion from the auto-completion menu, then pressing `<Tab>` will not get you to the next placeholder, but instead enter a `<Tab>` character, which is supposed to be the fallback. If you type out the whole suggestion yourself instead of accepting it from `nvim-cmp`, then it does go to the next placeholder.

Does this have something to do with `nvim-snippets` cmp source? Or is it inherent to how native snippets are implemented in Neovim?

**Reply by folke** [reply]
I think that's probably an issue with vim.snippet. it sets some events on which it removes the placeholders. I'm still planning to look into this. Might also be something we're doing wrong, but I think not

**Reply by folke** [reply]
I found the issue. The problem is that cmp **always** calls snippet.expand, even for items that are not a snippet.
This means that when you confirm and item from the new completions for the first placeholder, the snippet expansion cancels.

Will fix.

**Reply by folke** [reply]
Should be fixed now! https://github.com/LazyVim/LazyVim/commit/53911748ddc7f92793c6fec6348a9cde32159ef3

**Reply by dpetka2001** [reply]
Thank you very much. Can confirm it works now at least on my end. :smiley:

**Comment by nipuna-perera** [comment]
I noticed this too. I couldn't get tab to work well with even basic
snippets. It jumps once but then just opens the cmp menu instead of jumping
to next.
